atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/ring/RingGraph.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.workload.workflow.ring;
+
+import com.google.common.collect.Iterators;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import one.util.streamex.EntryStream;
+import one.util.streamex.IntStreamEx;
+
+public final class RingGraph {
+    private final Map<Integer, Optional<Integer>> ring;
+
+    private RingGraph(Map<Integer, Optional<Integer>> ring) {
+        this.ring = ring;
+    }
+
+    public static RingGraph init(int size) {
+        return new RingGraph(IntStreamEx.range(size)
+                .boxed()
+                .mapToEntry(_value -> Optional.<Integer>empty())
+                .toMap());
+    }
+
+    public static RingGraph from(Map<Integer, Integer> ring) {
+        return new RingGraph(EntryStream.of(ring).mapValues(Optional::of).toMap());
+    }
+
+    public static RingGraph fromPartial(Map<Integer, Optional<Integer>> ring) {
+        return new RingGraph(ring);
+    }
+
```

Comment:
So, I spent some time playing with this and I think using `Map` and `Optional` here are redundant. We could make the code maybe simpler if we use a `List` (or even `int[]`)

What I am thinking of is using a list kind of as a graph.

- The vertices are integers and `index -> list[index]` is an edge.
- `list[index] == index` means the node is isolated
- In order to validate that the graph is a ring, we'd need to do the following (maybe all in a loop)
  - Check all values in the list are >= 0 and < length(list) (otherwise data corruption TM)
  - Check no value repeats (otherwise, missingEntries)
  - Start at `i=0` and loop with `i = list[i]` until those two are the same. If the loop, runs exactly `length(list)` times, you know you have a ring (otherwise, found a smaller cycle).
  
  I quickly wrote something and it seemed lighter at least to my eyes. Also, you get to remove `fromPartial` and encode the num

Reply:
(as an aside, I think this is what is typically used when implementing disjoint-sets/union find; or graphs where you know your vertex set is statically in an integer range)

Reply:
Actually, do we even need anything custom at all here?

For a given ring size `n`, a ring can basically be thought of as a permutation of the list `0,..,n-1`

To generate a ring, we can just shuffle that list. And, when we read it from the store, we can read it as a map. Perform the checks there; and if valid output the permutation (one method?)

Anyway, just thinking out loud here

Reply:
oh yeah list would make this much prettier! 

I left it as a `Map<Integer, Optional<Integer>>` as that is the return type from the KVS, and there are some quirks with just using a `int[]` and `list` (you make the implicit assumption that nodeId is always the ordinalIndex; you also need to validate that each edge does not exceed the bounds of the array, which could be valid given that node ids may not be ordinal indexes); I'd push back against putting the validation logic in the workflow class itself to maintain separation of concerns. The workflow class should just be reading/writing to the KVS, and the state/representation of it's data can be validated by separate class (+ makes testing far more easier!).

Reply:
also one other thing I realized after our conversation: `To generate a ring, we can just shuffle that list` would not be possible, as if we point to node `0` in position `0`, then we generate a broken ring :/

Reply:
And that `List` does not guarantee uniqueness of it's value; we could check, but rather that we just have a data type that enforces this, such as `Set` or `Map`